# XML, HTML, and Web Scraping

JSON and XML are two different ways to represent hierarchical data. Which one is better? There are lots of articles online which discuss similarities and differences between JSON and XML and their advantages and disadvantages. Both formats are still in current usage, so it is good to be familiar with both. However, JSON is more common, so we'll focus on working with JSON representations of hierarchical data.

The reading covered an example of using Beautiful Soup to parse XML. Rather than doing another example XML now, we'll skip straight to scraping HTML from a webpage. Both HTML and XML can be parsed in a similar way with Beautiful Soup.

In [ ]:
import pandas as pd

## Scraping an HTML table with Beautiful Soup

Open the URL https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population and scroll down until you see a table of the cities in the U.S. with population over 100,000 (as of Jul 1, 2022). We'll use Beautiful Soup to scrape information from this table.

Read in the HTML from the ULR using the `requests` library.

In [ ]:
# YOUR CODE HERE
import requests
response = requests.get("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")

Use Beautiful Soup to parse this string into a tree called `soup`

In [ ]:
# YOUR CODE HERE
from bs4 import BeautifulSoup
soup = BeautifulSoup(response.content, "html.parser")

To find an HTML tag corresponding to a specific element on a webpage, right-click on it and choose "Inspect element". Go to the cities table Wikipedia page and do this now.

You should find that the cities table on the Wikipedia page corresponds to the element

```
<table class="wikitable sortable jquery-tablesorter" style="text-align:center">
```

There are many `<table>` tags on the page.

In [ ]:
len(soup.find_all("table"))

14

We can use attributes like `class=` and `style=` to narrow down the list.

In [ ]:
len(soup.find_all("table",
                  attrs={
                      "class": "wikitable sortable",
                      "style": "text-align:center"}
                  ))
# there are 3 tables that have the attributes for class and style

3

At this point, you can manually inspect the tables on the webpage to find that the one we want is the first one (see `[0]` below). We'll store this as `table`.

In [ ]:
table = soup.find_all("table",
                  attrs={
                      "class": "wikitable sortable",
                      "style": "text-align:center"}
                  )[0]
# here we are looking for the first table with this specifc class and style
table

**Now you will write code to scrape the information in `table` to create a Pandas data frame with one row for each city and columns for: city, state, population (2022 estimate), and 2020 land area (sq mi).** Refer to the Notes/suggestions below as you write your code. A few Hints are provided further down, but try coding first before looking at the hints.

Notes/suggestions:

- Use as a guide the code from the reading that produced the data frame of Statistics faculty
- Inspect the page source as you write your code
- You will need to write a loop to get the information for all cities, but you might want to try just scraping the info for New York first
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `.get_text(strip = True)` instead of `.text`
- Don't forget to convert to a Pandas Data Frame; it should have 333 rows and 4 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it --- e.g., what is the population density for all cities in CA? --- then you would need to clean the data first (to clean strings and convert to quantitative). (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)

In [ ]:
# YOUR CODE HERE. ADD AS MANY CELLS AS NEEDED

In [ ]:
# initialize an empty list
rows = []

# iterate over all rows in the faculty table
for city in table.find_all("tr")[1:]:

    # Get all the cells (<td>) in the row.
    cells = city.find_all("td")

    # The information we need is the text between tags.
    city_tag = cells[0].find("a") or cells[0]
    city = city_tag.text

    state_tag = cells[1].find("a") or cells[1]
    state = state_tag.get_text(strip = True)

    population_tag = cells[2].find("td") or cells[2]
    population = population_tag.get_text(strip = True)

    land_area_tag = cells[5].find("td") or cells[5]
    land_area = land_area_tag.get_text(strip = True)


    # Append this data.
    rows.append({
        "city": city,
        "state": state,
        "population": population,
        "land area": land_area

    })

rows

In [ ]:
pd.DataFrame(rows) # make it into a pandas dataframe

city          state population    land area
0       New York       New York  8,335,897  300.5 sq mi
1    Los Angeles     California  3,822,238  469.5 sq mi
2        Chicago       Illinois  2,665,039  227.7 sq mi
3        Houston          Texas  2,302,878  640.4 sq mi
4        Phoenix        Arizona  1,644,409  518.0 sq mi
..           ...            ...        ...          ...
328       Albany       New York    100,826   21.4 sq mi
329     Hesperia     California    100,744   72.7 sq mi
330  New Bedford  Massachusetts    100,682   20.0 sq mi
331    Davenport           Iowa    100,486   63.8 sq mi
332    Daly City     California    100,007    7.6 sq mi

[333 rows x 4 columns]

Hints:

- Each city is a row in the table; find all the `<tr>` tags to find all the cities
- Look for the `<td>` tag to see table entries within a row
- The rank column is represented by `<th>` tags, rather than `<td>` tags. So within a row, the first (that is, `[0]`) `<td>` tag corresponds to the city name.

## Aside: Scraping an HTML table with Pandas



The Pandas command `read_html` can be used to scrape information from an HTML table on a webpage.

We can call `read_html` on the URL.

In [ ]:
import pandas as pd
pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")

However, this scrapes all the tables on the webpage, not just the one we want. As with Beautiful Soup, we can narrow the search by specifying the table attributes.

In [ ]:
results= pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population",
             attrs = {'class': 'wikitable sortable', "style": "text-align:center"})
len(results) # there are 3 tables that have these attributes

3

This still returns 3 tables. As we remarked above, the table that we want is the first one (see `[0]` below).

In [ ]:
df_cities2 = pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population",
                          attrs = {'class': 'wikitable sortable', "style": "text-align:center"})[0]
df_cities2 # now we only want the first table

Wait, that seemed much easier than using Beautiful Soup, and it returned a data frame, and we even got for free some formatting like removing the commas from the population! Why didn't we just use `read_html` in the first place? It's true the `read_html` works well when scraping information from an HTML *table*. Unfortunately, you often want to scrape information from a webpage that isn't conveniently stored in an HTML table, in which case `read_html` won't work. (It only searches for `<table>`, `<th>`, `<tr>`, and `<td>` tags, but there are many other HTML tags.) Though Beautiful Soup is not as simple as `read_html`, it is more flexible and thus more widely applicable.

## Scraping information that is NOT in a `<table>` with Beautiful Soup

The Cal Poly course catalog http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory contains a list of courses offered by the Statistics department. **You will scrape this website to obtain a Pandas data frame with one row for each DATA or STAT course and two columns: course name and number (e.g, DATA 301. Introduction to Data Science) and term typically offered (e.g., Term Typically Offered: F, W, SP).**

Note: Pandas `read_html` is not help here since the courses are not stored in a `<table>.`

In [ ]:
pd.read_html("http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory")
# data is not in a table format so we can't use the read HTML pandas function

[                                        Program name   Program type
 0                              Actuarial Preparation          Minor
 1  Cross Disciplinary Studies Minor in Bioinforma...          Minor
 2   Cross Disciplinary Studies Minor in Data Science          Minor
 3                                         Statistics  BS, MS, Minor]


Notes/suggestions:


- Inspect the page source as you write your code
- The courses are not stored in a `<table>`. How are they stored?
- You will need to write a loop to get the information for all courses, but you might want to try just scraping the info for DATA 100 first
- What kind of tag is the course name stored in? What is the `class` of the tag?
- What kind of tag is the quarter(s) the course is offered stored in? What is the `class` of the tag? Is this the only tag of this type with the class? How will you get the one you want?
- You don't have to remove the number of units (e.g., 4 units) from the course name and number, but you can try it if you want
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `get_text(strip = True)` instead of `text`
- Don't forget to convert to a Pandas Data Frame; it should have 74 rows and 2 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it then you might need to clean the data first. (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)



In [ ]:
# YOUR CODE HERE. ADD AS MANY CELLS AS NEEDED

In [ ]:
import requests
# get the HTML source code
response = requests.get("https://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/")

In [ ]:
from bs4 import BeautifulSoup
soup1 = BeautifulSoup(response.content, "html.parser")
soup1

In [ ]:
soup1.find_all("div", {"class": "courseblock"})[0]
# you don't need to scroll through the website you can just look at the first version

In [ ]:
# initialize an empty list
rows = []

# iterate over all rows in the faculty table
for course in soup.find_all("div", {"class": "courseblock"}):
  # iterating with course not soup

    # The information we need is the text between tags.
    # you need to make sure you you iterate with course not soup so you get all the values
    course_tag = course.find("p", {"class": "courseblocktitle"})
    course1 = course_tag.get_text(strip = True)

    term_tag = course.find("p", {"class": "noindent"})
    term2 = term_tag.get_text(strip = True)


    # Append this data.
    rows.append({
        "Course Offered": course1,
        "Term": term2

    })

pd.DataFrame(rows)

Course Offered  \
0            DATA 100. Data Science for All I.4 units   
1      DATA 301. Introduction to Data Science.4 units   
2   DATA 401. Data Science Process and Ethics.3 units   
3   DATA 402. Mathematical Foundations of Data Sci...   
4   DATA 403. Data Science Projects Laboratory.1 unit   
..                                                ...   
69      STAT 551. Statistical Learning with R.4 units   
70    STAT 566. Graduate Consulting Practicum.2 units   
71       STAT 570. Selected Advanced Topics.1-4 units   
72    STAT 590. Graduate Seminar in Statistics.1 unit   
73                         STAT 599. Thesis.1-4 units   

                                Term  
0   Term Typically Offered: F, W, SP  
1   Term Typically Offered: F, W, SP  
2          Term Typically Offered: F  
3          Term Typically Offered: F  
4          Term Typically Offered: F  
..                               ...  
69         Term Typically Offered: F  
70        Term Typically Offered: SP  
71       Term Typically Offered: TBD  
72  Term Typically Offered: F, W, SP  
73       Term Typically Offered: TBD  

[74 rows x 2 columns]

Hints:

- Each course is represented by a `<div>` with `class=courseblock`, so you can find all the courses with `soup.find_all("div", {"class": "courseblock"})`
- The course name is in a `<p>` tag with `class=courseblocktitle`, inside a `<strong>` tag. (Though I don't think we need to find the strong tag here.)
- The term typically offered is in `<p>` tag with `class=noindent`. However, there are several tags with this class; term typically offered is the first one.
- If you want to use Beautiful Soup to remove the course units (e.g., 4 units), find the `<span>` tag within the course name tag and `.extract()` this span tag